### PR TITLE
SQA documentation for stochastic tools and optimization module

### PR DESCRIPTION
**Pull request by zachmprince**
Refs #15968

### CODE REVIEW

File: modules/optimization/doc/content/modules/optimization/sqa/optimization_srs.md [hunk 1]
```diff
@@ -1 +1,18 @@
-!template load file=app_srs.md.template app=Optimization category=optimization
+!template load file=sqa/module_srs.md.template app=Optimization category=optimization
+
+!template! item key=system-purpose
+!! system-purpose-begin
+The purpose of the MOOSE Optimization module optimize model parameters such that it minimizes a defined objective. An example would be to search for material property values so that the solution matches experimental results. As such, the purpose of this module is not to provide physical model capabilities, which is typically the responsibility of other MOOSE modules and dependent applications.
```
Comment (lynnmunday) [inline modules/optimization/doc/content/modules/optimization/sqa/optimization_srs.md:5]: The purpose of the MOOSE Optimization module is to control model parameters

**Review by lynnmunday — CHANGES_REQUESTED** [review]

File: modules/optimization/doc/content/modules/optimization/sqa/optimization_sdd.md [hunk 2]
```diff
@@ -1 +1,60 @@
...
+!template-end!
+
+!template! item key=system-scope
+!include optimization_srs.md start=system-scope-begin end=system-scope-finish
+!template-end!
+
+!template! item key=dependencies-and-limitations
+The {{module}} module inherits the
+[software dependencies of the MOOSE framework](framework_sdd.md#dependencies-and-limitations),
+with no additional dependencies. The module relies heavily on the TAO optimization package, which is available through the framework's inclusion of the PETSc. The design of this module is motivated by the
```
Comment (cticenhour) [inline modules/optimization/doc/content/modules/optimization/sqa/optimization_sdd.md:16]:
```suggestion
with no additional dependencies. The module relies heavily on the TAO optimization package, which is available through the framework's inclusion of PETSc. The design of this module is motivated by the
```

File: modules/optimization/doc/content/modules/optimization/sqa/optimization_sdd.md [hunk 3]
```diff
@@ -1 +1,60 @@
...
+!template! item key=design-stakeholders
+!include framework_sdd.md start=design-stakeholders-begin end=design-stakeholders-finish
+!template-end!
+
+!template! item key=system-design
+The {{module}} module inherits the wide range of pluggable systems from MOOSE.
+More information regarding MOOSE system design can be found in the framework
+[framework_sdd.md#system-design] section. Most of these inherited systems are
+utilized to build physics models for which the optimization is being
+performed for. Other than these, the MultiApps, Transfers, and Reporters systems
```
Comment (cticenhour) [inline modules/optimization/doc/content/modules/optimization/sqa/optimization_sdd.md:29]:
```suggestion
utilized to build physics models for which optimization is being
performed. Other than these, the MultiApps, Transfers, and Reporters systems
```

Also might want to link to these system landing pages, but that's optional to me.

File: modules/optimization/doc/content/modules/optimization/sqa/optimization_sdd.md [hunk 4]
```diff
@@ -1 +1,60 @@
...
+The architecture of the {{module}} module consists of a core and several pluggable systems.
+The core of MOOSE consists of a number of key objects responsible for setting up and managing the
+user-defined objects of a finite element simulation. This core set of objects has limited
+extendability and exists for every simulation configuration that the module is capable of running.
+
+!syntax complete subsystems=False actions=False objects=False groups=StochasticToolsApp
+
+!style halign=left
+The MooseApp is the top-level object used to hold all of the other objects in a simulation. In a
+normal simulation a single MooseApp object is created and "run()". This object uses its Factory
```
Comment (cticenhour) [inline modules/optimization/doc/content/modules/optimization/sqa/optimization_sdd.md:48]:
```suggestion
normal simulation, a single MooseApp object is created and "run()". This object uses its Factory
```

File: modules/optimization/doc/content/modules/optimization/sqa/optimization_srs.md (hunk 1 above)
Comment (cticenhour) [reply modules/optimization/doc/content/modules/optimization/sqa/optimization_srs.md:5]: This should be improved in terms of grammar anyway - will leave it to you and Zach to come to a suitable solution on specific wording.

File: modules/optimization/doc/content/modules/optimization/sqa/optimization_sdd.md [hunk 5]
```diff
@@ -1 +1,60 @@
...
+executed. The Finite Element data is stored in the Systems and Assembly object while the domain
+information (the Mesh) is stored in the Mesh object. A series of threaded loops are used to run
+parallel calculations on the objects created and stored within the warehouses.
+
+MOOSE's pluggable systems are documented on the [MOOSE website](https://mooseframework.inl.gov), and those
+for the {{module}} module are on this webpage, accessible through the high-level system links above.
+Each of these systems has a set of defined polymorphic interfaces and are designed to accomplish a
+specific task within the simulation. The design of these systems is solid and is managed through agile
+methods and ticket request system either on GitHub (for MOOSE) or on the defined software repository
+for this application.
```
Comment (cticenhour) [inline modules/optimization/doc/content/modules/optimization/sqa/optimization_sdd.md:59]:
```suggestion
methods and ticket request system on the [MOOSE GitHub repository](https://github.com/idaholab/moose).
```

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_sdd.md [hunk 6]
```diff
@@ -1 +1,64 @@
...
+!template! item key=design-stakeholders
+!include framework_sdd.md start=design-stakeholders-begin end=design-stakeholders-finish
+!template-end!
+
+!template! item key=system-design
+The {{module}} module inherits the wide range of pluggable systems from MOOSE. More
+information regarding MOOSE system design can be found in the framework [framework_sdd.md#system-design]
+section.
+Most of these inherited systems are utilized to build physics models for which the stochastic analysis
+is being performed for. Other than these, the inherited core capabilities define the order and
```
Comment (cticenhour) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_sdd.md:29]:
```suggestion
Most of these inherited systems are utilized to build physics models for which stochastic analysis
is being performed. Other than these, the inherited core capabilities define the order and
```

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_sdd.md [hunk 7]
```diff
@@ -1 +1,64 @@
...
+frequency of object execution. Furthermore, the MultiApps, Transfers, Controls, and Reporters systems
```
Comment (cticenhour) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_sdd.md:30]: Again, maybe consider links to the index pages, but optional to me.

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_sdd.md [hunk 8]
```diff
@@ -1 +1,64 @@
...
+The architecture of the {{module}} module consists of a core and several pluggable systems.
+The core of MOOSE consists of a number of key objects responsible for setting up and managing the
+user-defined objects of a finite element simulation. This core set of objects has limited
+extendability and exists for every simulation configuration that the module is capable of running.
+
+!syntax complete subsystems=False actions=False objects=False groups=StochasticToolsApp
+
+!style halign=left
+The MooseApp is the top-level object used to hold all of the other objects in a simulation. In a
+normal simulation a single MooseApp object is created and "run()". This object uses its Factory
```
Comment (cticenhour) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_sdd.md:52]:
```suggestion
normal simulation, a single MooseApp object is created and "run()". This object uses its Factory
```

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_sdd.md [hunk 9]
```diff
@@ -1 +1,64 @@
...
+executed. The Finite Element data is stored in the Systems and Assembly object while the domain
+information (the Mesh) is stored in the Mesh object. A series of threaded loops are used to run
+parallel calculations on the objects created and stored within the warehouses.
+
+MOOSE's pluggable systems are documented on the [MOOSE website](https://mooseframework.inl.gov), and those
+for the {{module}} module are on this webpage, accessible through the high-level system links above.
+Each of these systems has a set of defined polymorphic interfaces and are designed to accomplish a
+specific task within the simulation. The design of these systems is solid and is managed through agile
+methods and ticket request system either on GitHub (for MOOSE) or on the defined software repository
+for this application.
```
Comment (cticenhour) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_sdd.md:63]:
```suggestion
methods and ticket request system on the [MOOSE GitHub repository](https://github.com/idaholab/moose).
```

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_srs.md [hunk 10]
```diff
@@ -1 +1,48 @@
...
+As such, the purpose of this module is not to provide physical model
+capabilities, which is typically the responsibility of other MOOSE modules and
+dependent applications, but to provide data creation and processing capabilities
+in stochastic analysis.
+!! system-purpose-finish
+!template-end!
+
+!template! item key=system-scope
+!! system-scope-begin
+The MOOSE Stochastic Tools module builds on the existing framework MultiApps
```
Comment (cticenhour) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_srs.md:27]: Consider linking to MultiApps index.

File: modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_srs.md [hunk 11]
```diff
@@ -1 +1,48 @@
...
+system and provides several additional systems to address its purpose. The
+address the efficient evaluation of multiphysics models, the MultiApps systems
```
Comment (cticenhour) [inline modules/stochastic_tools/doc/content/modules/stochastic_tools/sqa/stochastic_tools_srs.md:29]:
```suggestion
system and provides several additional systems to address its purpose. To
address the efficient evaluation of multiphysics models, the MultiApps system
```

**Review by cticenhour — CHANGES_REQUESTED** [review]
This looks really good - a few minor comments and suggestions. Thanks!

**Review by grmnptr — CHANGES_REQUESTED** [review]
Nice!

**Review by cticenhour — CHANGES_REQUESTED** [review]
Double-checked the rendered SRS - still missing an assumptions and dependencies section and a reliability section for both of these modules.

**Comment by cticenhour** [comment]
@zachmprince looks like the current failures are a result of using `[#maintainability]` in the document to link to that SRS section. Not sure why it's not allowing you to link this - is it working fine locally?

**Comment by zachmprince** [comment]
> @zachmprince looks like the current failures are a result of using `[#maintainability]` in the document to link to that SRS section. Not sure why it's not allowing you to link this - is it working fine locally?

Took me way too long to debug, but I fixed it.

**Review by cticenhour — APPROVED** [review]
My comments have been addressed. Pending any other failures, I approve. @grmnptr do you have any other comments?

**Review by grmnptr — APPROVED** [review]

**Comment by cticenhour** [comment]
@zachmprince @grmnptr Will merge this tomorrow alongside some other SQA changes.

**Pull request merged**